File: Part1._Enumeration_problem/Lesson3._Backtracking_algorithm/Combination.c
#include <stdio.h>

int n = 0, k = 0;
int x[31];
FILE *fi = NULL, *fo = NULL;

void printResult()
{
    fprintf(fo, "{");
    for (int i = 1; i < k; i++)
    {
        fprintf(fo, "%d, ", x[i]);
    }
    fprintf(fo, "%d}\n", x[k]);
}

void Try(int i)
{
    for (int j = x[i - 1] + 1; j <= n - k + i; j++)
    {
        x[i] = j;
        if (i == k)
        {
            printResult();
        }
        else
        {
            Try(i + 1);
        }
    }
}

int main()
{
    fi = fopen("SUBSET.INP", "r");
    if (fi != NULL)
    {
        fscanf(fi, "%d %d", &n, &k);
        fclose(fi);
        if (k > 0 && k <= n && n <= 30)
        {
            fo = fopen("SUBSET.OUT", "w");
            x[0] = 0;
            Try(1);
            fclose(fo);
        }
    }
    return 0;
}

File: Part1._Enumeration_problem/Lesson3._Backtracking_algorithm/Combination.pas
program Combination;
const
    InputFile = 'SUBSET.INP';
    OutputFile = 'SUBSET.OUT';
    max = 30;
var
    x: array[0..max] of Integer;
    n, k: Integer;
    f: Text;

procedure PrintResult; (*In ra tập con {x[1], x[2], …, x[k]}*)
var
    i: Integer;
begin
    Write(f, '{');
    for i := 1 to k - 1 do Write(f, x[i], ', ');
    WriteLn(f, x[k], '}');
end;

procedure Try(i: Integer); {Thử các cách chọn giá trị cho x[i]}
var
    j: Integer;
begin
    for j := x[i - 1] + 1 to n - k + i do
        begin
            x[i] := j;
            if i = k then PrintResult
            else Try(i + 1);
        end;
end;

begin
    Assign(f, InputFile); Reset(F);
    ReadLn(f, n, k);
    Close(f);
    Assign(f, OutputFile); Rewrite(f);
    x[0] := 0;
    Try(1);
    Close(f);
end.